### ISSUE TITLE
Expose XmlTextReaderImpl.Normalization

### ISSUE THREAD

**Issue by ladipro** [dotnet/runtime#49418]
## Background and Motivation

`XmlTextReader` exposes the `Normalization` property, by default false, which affects whitespace and attribute value handling. No such property is exposed on `XmlReader` even though it is internally available on `XmlTextReaderImpl` where it is effectively hard-coded to true. If the developer wishes to move to the newer API, to be able to specify `XmlReaderSettings` for example, they have no choice but also change whitespace and attribute value handling, which may not be desirable.

## Proposed API

Technically the following should work but it may as well be a new property on `XmlReaderSettings`.

```diff
namespace System.Xml
{
    public abstract class XmlReader : IDisposable {
+        public virtual bool Normalization { get; set; }
     }
}
```

## Usage Examples

See https://github.com/dotnet/msbuild/pull/6232 for an example. The change would make XML parsing in MSBuild significantly faster if the property proposed here was available. In the PR `XmlTextReaderImpl.Normalization` is called via reflection as a proof of the concept.

## Alternative Designs

Unaware of alternative approaches other than implementing our own XML parser.

## Risks

N/A

**Comment by krwq** [comment]
@ladipro can you provide code samples and output for reference with what's currently not possible and what Normalization allows you to do?

**Comment by ladipro** [comment]
@krwq here's an example, the XML is copied from an MSBuild unit test:

```C#
string xml = @"<Project>
  <Target Name=""MultilineExec"">
    <Exec Command=""echo line 1
echo line 2
echo line 3"" />
  </Target>
</Project>";

XmlReader reader = XmlReader.Create(new StringReader(xml));

// HACK: Set Normalization to false to behave the same as XmlTextReader.
BindingFlags bindingFlags = BindingFlags.NonPublic | BindingFlags.SetProperty | BindingFlags.Instance;
reader.GetType().GetProperty("Normalization", bindingFlags).SetValue(reader, false);

XmlDocument document = new XmlDocument();
document.Load(reader);

Console.WriteLine(document.DocumentElement.FirstChild.FirstChild.Attributes["Command"].Value);
```

Without setting `Normalization` to false, newlines in the attribute value are converted to spaces and the code prints "echo line 1 echo line 2 echo line 3". We must be backward compatible and there really are scenarios where developers put newlines in attribute values which we have to preserve.

The switch from `XmlTextReader` to `XmlReader` is motivated by performance. Specifically, in many cases we can use `XmlReaderSettings` with `IgnoreComments` and `IgnoreWhitespace` which speeds up parsing.

**Comment by krwq** [comment]
@ladipro couple of questions:

- would it make sense to add property on the XmlReaderSettings instead?
- would it make sense to make it an enum instead?
- will allow implementations of XmlReader be able to respect it?
- `Normalization` is fairly vague word, how would you define the behavior? Would it be only scoped to behavior in attribute values or also some other places?
- assuming we go ahead with API review process and it is approved would you be interested in implementing it?

The reason I'm asking is in order to add this property we need to:
- prepare API review request
- have a good justification for this API
- API has to be well defined in what it's doing
- we need to understand how much work would it be to fully support it

**Comment by ladipro** [comment]
@krwq thank you, I can try answering some of these but an area owner would probably be in better position to, for example, come up with a fitting name.

> would it make sense to add property on the XmlReaderSettings instead?

As long as it allows me to enable the behavior behind the (vaguely named for sure!) `Normalization` property, as [already exposed](https://docs.microsoft.com/en-us/dotnet/api/system.xml.xmltextreader.normalization?view=net-5.0) on `XmlTextReader` and internally available on `XmlTextReaderImpl`, I don't care if it's through `XmlReaderSettings` or `XmlReader` itself. I don't need to change it after the reader has been created, although it is possible that someone else might have such a requirement.

> would it make sense to make it an enum instead?

I guess it wouldn't hurt but I don't know how likely it is that we'll have a need for more than two options in the future.

> will all implementations of XmlReader be able to respect it?

My guess would be that if we expose it on `XmlReaderSettings`, it would be fine to respect it only on those System.Xml implementations where it makes sense, and ignore it in the rest.

> `Normalization` is fairly vague word, how would you define the behavior? Would it be only scoped to behavior in attribute values or also some other places?

For what it's worth, the behavior is defined here: https://docs.microsoft.com/en-us/dotnet/api/system.xml.xmltextreader.normalization?view=net-5.0.

> assuming we go ahead with API review process and it is approved would you be interested in implementing it?

Yes, I can donate some time towards it. The implementation should be trivial. Again, all I'm asking is for a public surface hole to be plugged. Currently developers who need to parse XML with "Normalization" disabled cannot migrate to `XmlReader.Create`.

**Comment by krwq** [comment]
Thanks @ladipro we'll talk about it during our next triage meeting and will decide if/how we will go forward with this.

**Comment by krwq** [comment]
Sorry for the wait, just wanted to let you know this is still on our radar.

**Comment by buyaa-n** [comment]
For me, it makes sense to add the property on the XmlReaderSettings instead and keep the name Normalization

**Comment by stephentoub** [comment]
@krwq, do you have an estimate for when you'll review this?

**Comment by joperezr** [comment]
> The switch from XmlTextReader to XmlReader is motivated by performance. Specifically, in many cases we can use XmlReaderSettings with IgnoreComments and IgnoreWhitespace which speeds up parsing.

[Triage]: Can you expand a bit more on why you think XmlReader is faster than XmlTextReader? XmlReader is an abstract class and the implementation you are getting underneath is XmlTextReader anyway, so in theory you should be abel to just use XmlTextReader and use the Normalization property it already has exposed. If the problem here is that XmlTextReader doesn't take XmlReaderSettings, then I think that should be what this proposal is about, so that you could still get the perf benefits that provides while still using XmlTextReader directly.

Main reason why we think this would be a better approach is due to the many different implementations of XmlReaders and the fact that maybe it would not be so easy to implement Normalization on all of them, and it seems like the one that you want already does support it.

**Comment by ladipro** [comment]
@joperezr yes, what you wrote is correct, making `XmlTextReader` take `XmlReaderSettings` would also do. It is a combination of the two - `XmlReaderSettings` (`IgnoreComments` and `IgnoreWhitespace` in particular) and `Normalization` - that we need, and if it's easier to make `XmlTextReader` accept `XmlReaderSettings` rather than make the newer API expose `Normalization`, then let's go for it.

**Comment by krwq** [comment]
It's definitely less work and is also useful for completeness. For new option in XmlReaderSettings we would need to investigate all XmlReader implementations to respect it and also potentially making a not supported option for people who implemented their own XmlReader. Scoping this to just XmlTextReader is much simpler and won't cause even more weird discrepancies related to whitespaces across all implementations

**Comment by krwq** [comment]
Next steps:
- prepare final API shape, so likely just XmlTextReader having new ctor
- investigate what is supported and if there are any limitation
- submit API review (simply replace label in this issue and wait)
- prepare PR with tests

and of course we need a volunteer for this work 😄

**Comment by joperezr** [comment]
Adding @dotnet/area-system-xml so they can help move this forward.
